### PR TITLE
SQL Server to Postgres Migration

### PR DESCRIPTION
**Pull request by labkey-martyp**
#### Rationale
1. age_in_days / age(..., SQL_TSI_DAY) — Preserves SQL Server's calendar-boundary semantics for timestampdiff(SQL_TSI_DAY, ...) on Postgres, where the native operator counts 24-hour intervals instead. Consistent with the existing age_in_months/age_in_years functions; permanent addition to LabKey SQL.
2. CAST(... AS BOOLEAN) in SELECT lists (QuerySelect.java) — On SQL Server, boolean QMethodCalls in a select list are wrapped in CASE WHEN (...) THEN 1 ELSE 0 END to avoid predicates as scalar values. But CAST(x AS BIT) is already a value, not a predicate, and SQL Server rejects the wrapped form. Fix excludes ConvertInfo from the wrap. (repro: insert failing query / error)
3. DatabaseMigrationService.registerSchemaContributor — Interface for DatabaseMigrationService.registerSchemaContributor in premiumModules

#### Related Pull Requests
- https://github.com/LabKey/premiumModules/pull/499
- https://github.com/LabKey/tnprcEHRModules/pull/277
- https://github.com/LabKey/ehrModules/pull/1119

#### Changes
- Method.java — age_in_days method, age(..., SQL_TSI_DAY) dispatch, updated validation
- QueryTestCase.jsp — tests for both new syntaxes including datetime calendar-boundary cases
- LabKeySql.md — documents age_in_days
- QuerySelect.java — skip boolean wrap for CAST/CONVERT on SQL Server
- DatabaseMigrationService.java — new registerSchemaContributor interface method

### CODE REVIEW

File: query/src/org/labkey/query/QueryTestCase.jsp [hunk 1]
```diff
@@ -777,6 +777,32 @@ d,seven,twelve,day,month,date,duration,guid
...
+        new MethodSqlTest("SELECT TIMESTAMPDIFF2(SQL_TSI_QUARTER, CAST('15 Oct 2003' AS TIMESTAMP), CAST('14 Jan 2003' AS TIMESTAMP))", JdbcType.INTEGER, -3),
+        new MethodSqlTest("SELECT TIMESTAMPDIFF2(SQL_TSI_YEAR, CAST('01 Jan 2003' AS TIMESTAMP), CAST('01 Jan 2006' AS TIMESTAMP))", JdbcType.INTEGER, 3),
+        new MethodSqlTest("SELECT TIMESTAMPDIFF2(SQL_TSI_YEAR, CAST('15 Jan 2003' AS TIMESTAMP), CAST('14 Jan 2006' AS TIMESTAMP))", JdbcType.INTEGER, 2),
+        new MethodSqlTest("SELECT TIMESTAMPDIFF2(SQL_TSI_YEAR, CAST('15 Jan 2003' AS TIMESTAMP), CAST('15 Jan 2006' AS TIMESTAMP))", JdbcType.INTEGER, 3),
+        new MethodSqlTest("SELECT TIMESTAMPDIFF2(SQL_TSI_YEAR, CAST('14 Jan 2006' AS TIMESTAMP), CAST('15 Jan 2003' AS TIMESTAMP))", JdbcType.INTEGER, -2),
+        new MethodSqlTest("SELECT TIMESTAMPDIFF2(SQL_TSI_FRAC_SECOND, CAST('01 Jan 2004 5:00:00' AS TIMESTAMP), CAST('01 Jan 2004 5:00:01' AS TIMESTAMP))", JdbcType.BIGINT, 1000),
+        new MethodSqlTest("SELECT TIMESTAMPDIFF2(SQL_TSI_FRAC_SECOND, CAST('01 Jan 2004' AS TIMESTAMP), CAST('31 Jan 2004' AS TIMESTAMP))", JdbcType.BIGINT, 2592000000L),
```
Comment (labkey-adam) [inline query/src/org/labkey/query/QueryTestCase.jsp]: Add test for `age_in_days()`?

Comment (labkey-adam) [reply query/src/org/labkey/query/QueryTestCase.jsp]: And new `age()` option?

File: api/src/org/labkey/api/data/dialect/BasePostgreSqlDialect.java [hunk 2]
```diff
@@ -1157,6 +1159,53 @@ private SQLFragment timestampdiff(SQLFragment... arguments)
...
+            case "SQL_TSI_DAY" ->
+                new SQLFragment("TRUNC(EXTRACT(EPOCH FROM (").append(end).append(") - (").append(start).append(")) / 86400)::INT");
+            case "SQL_TSI_HOUR" ->
+                new SQLFragment("TRUNC(EXTRACT(EPOCH FROM (").append(end).append(") - (").append(start).append(")) / 3600)::INT");
+            case "SQL_TSI_MINUTE" ->
+                new SQLFragment("TRUNC(EXTRACT(EPOCH FROM (").append(end).append(") - (").append(start).append(")) / 60)::INT");
+            case "SQL_TSI_SECOND" ->
+                new SQLFragment("TRUNC(EXTRACT(EPOCH FROM (").append(end).append(") - (").append(start).append(")))::INT");
+            case "SQL_TSI_FRAC_SECOND" ->
+                new SQLFragment("TRUNC(EXTRACT(EPOCH FROM (").append(end).append(") - (").append(start).append(")) * 1000)::BIGINT");
```
Comment (labkey-adam) [inline api/src/org/labkey/api/data/dialect/BasePostgreSqlDialect.java]: Claude is complaining about this `BIGINT` vs. `Method` defining `timestampdiff2` as returning `JdbcType.INTEGER`.

File: api/src/org/labkey/api/data/dialect/BasePostgreSqlDialect.java [hunk 3]
```diff
@@ -1157,6 +1159,53 @@ private SQLFragment timestampdiff(SQLFragment... arguments)
...
+            default -> throw new IllegalArgumentException("Unsupported interval for timestampdiff2: " + interval);
+        };
+    }
+
+    private SQLFragment getWholeElapsedMonths(SQLFragment start, SQLFragment end)
+    {
```
Comment (labkey-adam) [inline api/src/org/labkey/api/data/dialect/BasePostgreSqlDialect.java]: This method assumes that `start` and `end` don't include any parameters. Current callers don't, but best to add a comment and an invariant check for that, i.e., throw if either has parameters.

File: api/src/org/labkey/api/data/dialect/BasePostgreSqlDialect.java (hunk 2 above)
Comment (labkey-adam) [reply api/src/org/labkey/api/data/dialect/BasePostgreSqlDialect.java]: In limited local testing, I didn't see any problem with very large values, though.

**Review by labkey-adam — CHANGES_REQUESTED** [review]
Please review suggestions from me and Claude. I also suggest running Claude /review_pr after changes. If its findings are off, consider adding comments that address.

File: query/src/org/labkey/query/QueryTestCase.jsp (hunk 1 above)
Comment (labkey-martyp) [reply query/src/org/labkey/query/QueryTestCase.jsp]: added

**Comment by labkey-martyp** [comment]
@labkey-adam After a little further testing, it turns out we don't really need timestampdiff2 for this migration. age_in_days doesn't have the parameter swapping issue and will work for all use cases. I also added tests for age_in_days. /review-pr is just returning a few low severity findings.

**Review by labkey-adam — APPROVED** [review]
